scipy/_lib/doccer.py

```diff
@@ -202,73 +265,106 @@ def indentcount_lines(lines):
-        If True, strip common indentation from all parameters in
-        docdict
+    docdict : dict[str, str]
+        A dictionary containing name, docstring fragment pairs.
+    unindent_params : bool, optional
+        If True, strip common indentation from all parameters in docdict.
+        Default is False.
 
     Returns
     -------
-    decfunc : function
-        decorator that applies dictionary to input function docstring
-
-    '''
+    Decorator[R]
+        The decorator function that applies dictionary to its
+        argument's __doc__ attribute.
+    """
     if unindent_params:
         docdict = unindent_dict(docdict)
 
-    def decorate(f):
-        f.__doc__ = docformat(f.__doc__, docdict)
-        return f
+    def decorate(func: Callable[P, R]) -> Callable[P, R]:
+        # __doc__ may be None for optimized Python (-OO)
+        doc = func.__doc__ or ""
+        func.__doc__ = docformat(doc, docdict)
+        return func
+
     return decorate
 
 
-def unindent_dict(docdict):
-    ''' Unindent all strings in a docdict '''
-    can_dict = {}
+def unindent_dict(docdict: dict[str, str]) -> dict[str, str]:
+    """Unindent all strings in a docdict.
+
+    Parameters
+    ----------
+    docdict : dict[str, str]
+        A dictionary with string values to unindent.
+
+    Returns
+    -------
+    dict[str, str]
+        The `docdict` dictionary but each of its string values are unindented.
+    """
+    can_dict: dict[str, str] = {}
     for name, dstr in docdict.items():
         can_dict[name] = unindent_string(dstr)
     return can_dict
 
 
-def unindent_string(docstring):
-    ''' Set docstring to minimum indent for all lines, including first
+def unindent_string(docstring: str) -> str:
+    """Set docstring to minimum indent for all lines, including first.
+
+    Parameters
+    ----------
+    docstring : str
+        The input docstring to unindent.
+
+    Returns
+    -------
+    str
+        The unindented docstring.
 
+    Examples
+    --------
     >>> unindent_string(' two')
     'two'
     >>> unindent_string('  two\\n   three')
     'two\\n three'
-    '''
+    """
     lines = docstring.expandtabs().splitlines()
     icount = indentcount_lines(lines)
     if icount == 0:
         return docstring
-    return '\n'.join([line[icount:] for line in lines])
+    return "\n".join([line[icount:] for line in lines])
 
 
-def doc_replace(obj, oldval, newval):
+def doc_replace(obj: type[Any] | Any, oldval: str, newval: str) -> Decorator[R]:
     """Decorator to take the docstring from obj, with oldval replaced by newval
 
     Equivalent to ``func.__doc__ = obj.__doc__.replace(oldval, newval)``
 
     Parameters
     ----------
-    obj : object
-        The object to take the docstring from.
-    oldval : string
-        The string to replace from the original docstring.
-    newval : string
-        The string to replace ``oldval`` with.
+    obj : type[Any] | Any
```

Comment:
This kind of change is typically undesirable. Docstrings are first of all for humans to read; the goal is to keep this correct and readable, not to duplicate the annotations from the signature. The old `object` was file here.

Reply:
Thanks for the feedback. I have changed all of these types into just object in the docstrings.